src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1885,139 +2021,186 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
             }
 
-            ReadOnlySpan<uint> x = left.ToUInt32Span(stackalloc uint[StackallocUInt32Limit / 2]);
-            ReadOnlySpan<uint> y = right.ToUInt32Span(stackalloc uint[StackallocUInt32Limit / 2]);
-            uint[] z = new uint[Math.Max(x.Length, y.Length)];
             uint xExtend = (left._sign < 0) ? uint.MaxValue : 0;
             uint yExtend = (right._sign < 0) ? uint.MaxValue : 0;
 
+            uint[]? leftBufferFromPool = null;
+            int size = (left._bits?.Length ?? 1) + 1;
+            Span<uint> x = ((uint)size <= BigIntegerCalculator.StackAllocThreshold ?
+                           stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                           : leftBufferFromPool = ArrayPool<uint>.Shared.Rent(size)).Slice(0, size);
+            x = x.Slice(0, left.WriteTo(x));
+
+            uint[]? rightBufferFromPool = null;
+            size = (right._bits?.Length ?? 1) + 1;
+            Span<uint> y = ((uint)size <= BigIntegerCalculator.StackAllocThreshold ?
+                           stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                           : rightBufferFromPool = ArrayPool<uint>.Shared.Rent(size)).Slice(0, size);
+            y = y.Slice(0, right.WriteTo(y));
+
+            uint[]? resultBufferFromPool = null;
+            size = Math.Max(x.Length, y.Length);
+            Span<uint> z = (size <= BigIntegerCalculator.StackAllocThreshold ?
+                           stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                           : resultBufferFromPool = ArrayPool<uint>.Shared.Rent(size)).Slice(0, size);
+
             for (int i = 0; i < z.Length; i++)
             {
-                uint xu = (i < x.Length) ? x[i] : xExtend;
-                uint yu = (i < y.Length) ? y[i] : yExtend;
+                uint xu = ((uint)i < (uint)x.Length) ? x[i] : xExtend;
+                uint yu = ((uint)i < (uint)y.Length) ? y[i] : yExtend;
                 z[i] = xu ^ yu;
             }
 
-            return new BigInteger(z);
+            if (leftBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(leftBufferFromPool);
+
+            if (rightBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(rightBufferFromPool);
+
+            var result = new BigInteger(z);
+
+            if (resultBufferFromPool != null)
+                ArrayPool<uint>.Shared.Return(resultBufferFromPool);
+
+            return result;
         }
 
         public static BigInteger operator <<(BigInteger value, int shift)
         {
-            if (shift == 0)
-                return value;
+            if (shift == 0) return value;
+            else if (shift == int.MinValue) return ((value >> int.MaxValue) >> 1);
+            else if (shift < 0) return value >> -shift;
 
-            if (shift == int.MinValue)
-                return (value >> int.MaxValue) >> 1;
+            int digitShift = shift / kcbitUint;
```

Comment:
Why the move away from `Math.DivRem`? It should have already been producing optimal codegen.
(Same comment line 2184 below.)

Reply:
That the result of manual merge. I'll investigate that one more time.

Reply:
Reviewed and fixed: left and right shift operators.